### ISSUE TITLE
[Bug] Komga fails to close PDF files, may cause OOM

### ISSUE THREAD

**Issue by svew** [gotson/komga#784]
### Komga environment
- **Komga version**:
- [x] **I am running Komga with Docker**
  - Docker image tag [e.g. latest, beta]: 5f5804bbebbf
- [ ] **I am running Komga from the `jar`**
  - Java version:
- [ ] **I have a problem in the web interface**
  - Browser (with version):
- [ ] **I have a problem with an OPDS client application**
  - OPDS Application (with version):
- [ ] **I have a problem with the Tachiyomi extension**
  - Tachiyomi version:
  - Tachiyomi extension version:

### Describe the bug
My log files are full of this warning while analyzing files:

```
2022-01-27 13:00:23.493  WARN 1 --- [Finalizer] org.apache.pdfbox.cos.COSDocument        : Warning: You did not close a PDF Document
```

Which seem to lead to this warning:

```
2022-01-27 17:49:38.937  WARN 1 --- [org.springframework.jms.JmsListenerEndpointContainer#1-3] o.g.k.infrastructure.jms.ArtemisConfig   : Java heap space
```

And subsequent crashing due to being out of memory. Unsure if they're actually connected, so just reporting the failure to close PDF file

### Steps to reproduce
1. Add many PDF documents to library
2. Scan the library for new documents, check the logs
3. See the failure to close PDF files, and if enough PDFs, see increased memory

### Expected behavior
No warning log

### Actual behavior
The above warning log

### Log file
```
2022-01-27 17:49:32.161  WARN 1 --- [Finalizer] org.apache.pdfbox.cos.COSDocument        : Warning: You did not close a PDF Document
2022-01-27 17:49:32.161  WARN 1 --- [org.springframework.jms.JmsListenerEndpointContainer#2-5] o.s.j.l.DefaultMessageListenerContainer  : Setup of JMS message listener invoker failed for destination 'sse' - trying to recover. Cause: Java heap space
2022-01-27 17:49:32.252  WARN 1 --- [org.springframework.jms.JmsListenerEndpointContainer#1-3] o.g.k.infrastructure.jms.ArtemisConfig   : Java heap space
2022-01-27 17:49:32.256  INFO 1 --- [org.springframework.jms.JmsListenerEndpointContainer#1-3] o.g.komga.application.tasks.TaskHandler  : Executing task: AnalyzeBook(bookId='07KYJJR9VC2XK', priority='4')
2022-01-27 17:49:32.257  INFO 1 --- [org.springframework.jms.JmsListenerEndpointContainer#1-3] o.g.komga.domain.service.BookLifecycle   : Analyze and persist book: Book(name=JJBA 5 - Stone Ocean 13, url=file:/data/Manga/JJBA%206%20-%20Stone%20Ocean/JJBA%205%20-%20Stone%20Ocean%2013.pdf, fileLastModified=2014-12-27T16:21:11, fileSize=165094497, fileHash=131jxpc, number=13, id=07KYJJR9VC2XK, seriesId=07KYJJR9QC4FD, libraryId=07KYHZ1G3C8VS, deletedDate=null, createdDate=2022-01-27T14:39:09, lastModifiedDate=2022-01-27T17:46:42.696)
2022-01-27 17:49:32.258  INFO 1 --- [org.springframework.jms.JmsListenerEndpointContainer#1-3] o.g.komga.domain.service.BookAnalyzer    : Trying to analyze book: Book(name=JJBA 5 - Stone Ocean 13, url=file:/data/Manga/JJBA%206%20-%20Stone%20Ocean/JJBA%205%20-%20Stone%20Ocean%2013.pdf, fileLastModified=2014-12-27T16:21:11, fileSize=165094497, fileHash=131jxpc, number=13, id=07KYJJR9VC2XK, seriesId=07KYJJR9QC4FD, libraryId=07KYHZ1G3C8VS, deletedDate=null, createdDate=2022-01-27T14:39:09, lastModifiedDate=2022-01-27T17:46:42.696)
2022-01-27 17:49:32.260  INFO 1 --- [org.springframework.jms.JmsListenerEndpointContainer#1-3] o.g.komga.domain.service.BookAnalyzer    : Detected media type: application/pdf
2022-01-27 17:49:38.850  WARN 1 --- [Finalizer] org.apache.pdfbox.cos.COSDocument        : Warning: You did not close a PDF Document
2022-01-27 17:49:38.936  WARN 1 --- [Thread-6 (ActiveMQ-server-org.apache.activemq.artemis.core.server.impl.ActiveMQServerImpl$6@30a62a5b)] org.apache.activemq.artemis.core.server  : AMQ222149: Message Reference[10737423979]:RELIABLE:CoreMessage[messageID=10737423979,durable=true,userID=262fc483-7fdc-11ec-8f52-0242ac110002,priority=4, timestamp=Thu Jan 27 17:46:45 PST 2022,expiration=0, durable=true, address=tasks.background,size=625,properties=TypedProperties[subtype=AnalyzeBook,_AMQ_GROUP_ID=D,__AMQ_CID=fe3f1351-7fd9-11ec-8f52-0242ac110002,unique_id=ANALYZE_BOOK_07KYJJR9VC2XK,_AMQ_ROUTING_TYPE=1,type=task]]@277474725 has reached maximum delivery attempts, sending it to Dead Letter Address DLQ from tasks.background
2022-01-27 17:49:38.937  WARN 1 --- [org.springframework.jms.JmsListenerEndpointContainer#1-3] o.g.k.infrastructure.jms.ArtemisConfig   : Java heap space
2022-01-27 17:49:38.941  INFO 1 --- [org.springframework.jms.JmsListenerEndpointContainer#1-3] o.g.komga.application.tasks.TaskHandler  : Executing task: AnalyzeBook(bookId='07KYJJR9VC2XM', priority='4')
2022-01-27 17:49:38.942  INFO 1 --- [org.springframework.jms.JmsListenerEndpointContainer#1-3] o.g.komga.domain.service.BookLifecycle   : Analyze and persist book: Book(name=JJBA 5 - Stone Ocean 14, url=file:/data/Manga/JJBA%206%20-%20Stone%20Ocean/JJBA%205%20-%20Stone%20Ocean%2014.pdf, fileLastModified=2014-12-27T16:22:34, fileSize=153702496, fileHash=py6jwd, number=14, id=07KYJJR9VC2XM, seriesId=07KYJJR9QC4FD, libraryId=07KYHZ1G3C8VS, deletedDate=null, createdDate=2022-01-27T14:39:09, lastModifiedDate=2022-01-27T17:46:42.697)
2022-01-27 17:49:38.944  INFO 1 --- [org.springframework.jms.JmsListenerEndpointContainer#1-3] o.g.komga.domain.service.BookAnalyzer    : Trying to analyze book: Book(name=JJBA 5 - Stone Ocean 14, url=file:/data/Manga/JJBA%206%20-%20Stone%20Ocean/JJBA%205%20-%20Stone%20Ocean%2014.pdf, fileLastModified=2014-12-27T16:22:34, fileSize=153702496, fileHash=py6jwd, number=14, id=07KYJJR9VC2XM, seriesId=07KYJJR9QC4FD, libraryId=07KYHZ1G3C8VS, deletedDate=null, createdDate=2022-01-27T14:39:09, lastModifiedDate=2022-01-27T17:46:42.697)
2022-01-27 17:49:39.205  INFO 1 --- [org.springframework.jms.JmsListenerEndpointContainer#1-3] o.g.komga.domain.service.BookAnalyzer    : Detected media type: application/pdf
2022-01-27 17:49:53.380  WARN 1 --- [Finalizer] org.apache.pdfbox.cos.COSDocument        : Warning: You did not close a PDF Document
2022-01-27 17:49:53.383  WARN 1 --- [org.springframework.jms.JmsListenerEndpointContainer#0-6] o.s.j.l.DefaultMessageListenerContainer  : Setup of JMS message listener invoker failed for destination 'sse' - trying to recover. Cause: Java heap space
2022-01-27 17:49:53.574  WARN 1 --- [org.springframework.jms.JmsListenerEndpointContainer#1-3] o.g.k.infrastructure.jms.ArtemisConfig   : Java heap space
2022-01-27 17:49:53.579  INFO 1 --- [org.springframework.jms.JmsListenerEndpointContainer#1-3] o.g.komga.application.tasks.TaskHandler  : Executing task: AnalyzeBook(bookId='07KYJJR9VC2XM', priority='4')
2022-01-27 17:49:53.580  INFO 1 --- [org.springframework.jms.JmsListenerEndpointContainer#1-3] o.g.komga.domain.service.BookLifecycle   : Analyze and persist book: Book(name=JJBA 5 - Stone Ocean 14, url=file:/data/Manga/JJBA%206%20-%20Stone%20Ocean/JJBA%205%20-%20Stone%20Ocean%2014.pdf, fileLastModified=2014-12-27T16:22:34, fileSize=153702496, fileHash=py6jwd, number=14, id=07KYJJR9VC2XM, seriesId=07KYJJR9QC4FD, libraryId=07KYHZ1G3C8VS, deletedDate=null, createdDate=2022-01-27T14:39:09, lastModifiedDate=2022-01-27T17:46:42.697)
2022-01-27 17:49:53.582  INFO 1 --- [org.springframework.jms.JmsListenerEndpointContainer#1-3] o.g.komga.domain.service.BookAnalyzer    : Trying to analyze book: Book(name=JJBA 5 - Stone Ocean 14, url=file:/data/Manga/JJBA%206%20-%20Stone%20Ocean/JJBA%205%20-%20Stone%20Ocean%2014.pdf, fileLastModified=2014-12-27T16:22:34, fileSize=153702496, fileHash=py6jwd, number=14, id=07KYJJR9VC2XM, seriesId=07KYJJR9QC4FD, libraryId=07KYHZ1G3C8VS, deletedDate=null, createdDate=2022-01-27T14:39:09, lastModifiedDate=2022-01-27T17:46:42.697)
2022-01-27 17:49:53.584  INFO 1 --- [org.springframework.jms.JmsListenerEndpointContainer#1-3] o.g.komga.domain.service.BookAnalyzer    : Detected media type: application/pdf
2022-01-27 17:50:01.762  WARN 1 --- [org.springframework.jms.JmsListenerEndpointContainer#1-3] o.g.k.infrastructure.jms.ArtemisConfig   : Java heap space
```

**Comment by svew** [comment]
Added a PR for a fix for this: #785

**Comment by svew** [comment]
Realizing this might not be the only issue with the scenario:

In my case, Komga is trying to analyze a series of 17 PDFs, which together have a size of 2.6GB. The Caffeine cache evicts entries based on time and number of entries, not entry size. Docker only seems to use up to about 2GB before Komga becomes unresponsive, so I'd assume my image's memory limit is around 2GB, but I only assume because I'm not sure how to check that limit.

Either way, it might be better to read the PDF file size, use it as the cache entry's weight, and set a `maximumWeight` for the cache.

**Comment by gotson** [comment]
Can you provide a complete DEBUG log file of the scenario?

**Comment by gotson** [comment]
> In my case, Komga is trying to analyze a series of 17 PDFs, which together have a size of 2.6GB.

PDFs have poor support in Komga, because Apache PDFBox requires to load the whole files in memory, even to get a single page. CBZ are recommended.

> The Caffeine cache evicts entries based on time and number of entries, not entry size.

The `PDDocument` doesn't represent the size of the PDF file.

> Docker only seems to use up to about 2GB before Komga becomes unresponsive, so I'd assume my image's memory limit is around 2GB, but I only assume because I'm not sure how to check that limit.

If you haven't set the memory limit of the JVM, it depends on the total memory of the host. Over 2gb of memory, the JVM will allocate 1/4th of the memory by default, but will grow as needed.

**Comment by gotson** [comment]
> Either way, it might be better to read the PDF file size, use it as the cache entry's weight, and set a `maximumWeight` for the cache.

That would effectively limit the loading of any PDF file whose size would be over the `maximumWeight`, which is probably not a good idea.

**Comment by svew** [comment]
That's true about the host memory, makes sense since the total memory I have on that system is 4GB

> That would effectively limit the loading of any PDF file whose size would be over the `maximumWeight`, which is probably not a good idea.

That would not be good behavior. The cache should exist to improve load times, not prevent loads from happening. But it's no worry, the [Caffeine cache already accounts for that](https://www.javadoc.io/doc/com.github.ben-manes.caffeine/caffeine/latest/com.github.benmanes.caffeine/com/github/benmanes/caffeine/cache/Caffeine.html#maximumWeight(@org.checkerframework.checker.index.qual.NonNegativelong)):

> Note that weight is only used to determine whether the cache is over capacity; it has no effect on selecting which entry should be evicted next. 

So a weight limit on the cache won't influence what files are allowed to load.

You're right that PDDocument doesn't really represent the file size, it could be larger, so checking the file size probably isn't the best option. I checked the [PDDocument docs](https://pdfbox.apache.org/docs/2.0.2/javadocs/org/apache/pdfbox/pdmodel/PDDocument.html#load(byte[])), and it looks like you can supply [MemoryUsageSettings](https://pdfbox.apache.org/docs/2.0.2/javadocs/org/apache/pdfbox/io/MemoryUsageSetting.html) which would allow you to restrict the amount of data you load into memory, and stores the rest in temporary files. Maybe we could add a configuration for maximum memory load size and for PDF cache size? It'd be a helpful thing to be able to modify depending on your system.

A better question though, why are we caching PDFs? I can't think of a scenario where you'd see a lot of cache hits...

**Comment by gotson** [comment]
> But it's no worry, the [Caffeine cache already accounts for that](https://www.javadoc.io/doc/com.github.ben-manes.caffeine/caffeine/latest/com.github.benmanes.caffeine/com/github/benmanes/caffeine/cache/Caffeine.html#maximumWeight(@org.checkerframework.checker.index.qual.NonNegativelong)):
> 
> > Note that weight is only used to determine whether the cache is over capacity; it has no effect on selecting which entry should be evicted next.
> 
> So a weight limit on the cache won't influence what files are allowed to load. If an entry with a weight higher than the max is loaded, it'll just clear the cache.

My understanding seems to be different that yours. What the doc says is that the weight is not used to decide which entry is removed first, this will still be decided by the policy, like last accessed, or last written. However, it doesn't say that the cache entry wouldn't be removed if the size is over the maximum. And it also says:

> Specifies the maximum weight of entries the cache may contain.

> When maximumWeight is zero, elements will be evicted immediately after being loaded into the cache.

IMH you won't be able to have a cache entry that is bigger than the `maximumSize`.

---



> Not quite sure what you meant when you said PDDocument doesn't represent the size of the PDF

What i am saying is the size of the `PDDocument` object may not exactly represent the size of the file on disk.



> I checked the [PDDocument docs](https://pdfbox.apache.org/docs/2.0.2/javadocs/org/apache/pdfbox/pdmodel/PDDocument.html#load(byte%5B%5D)), and it looks like you can supply [MemoryUsageSettings](https://pdfbox.apache.org/docs/2.0.2/javadocs/org/apache/pdfbox/io/MemoryUsageSetting.html) which would allow you to restrict the amount of data you load into memory, and stores the rest in temporary files.

That is interesting, i didn't know about that. I'm wondering how the use of temp files would help to reduce the memory footprint for PDFs.

> Maybe we could add a configuration for maximum memory load size and for PDF cache size? It'd be a helpful thing to be able to modify depending on your system.

I try to keep the extra knobs at a minimum, that usually only confuse people more. And we also need to provide good defaults that work most of the time.

> A better question though, why are we caching PDFs? I can't think of a scenario where you'd see a lot of cache hits...

Because when you read a book, you will read all pages one after the other, so if you request pages in sequence, every time you would reopen and reparse the PDF file entirely, which takes a long time. Just the fact of parsing the PDF to know what's inside could already be a couple of seconds.

**Comment by gotson** [comment]
I have tested to load PDFs using `PDDocument.load(path.toFile(), MemoryUsageSetting.setupTempFileOnly())`, this seems to be working.

I tried again to enable/disable the caching of `PDDocument`, and when cache is disabled the effect can be seen easily, for instance by opening the thumbnail view of a large pdf file. Enabling the cache would reduce the loading size by 1/3 at least.

Given everyone has been complaining about how PDF are a pain in Komga, and the memory usage, I would say it would be a sane default to use temp files for reading the PDF streams.

I don't think it makes much sense to provide a configuration option to use memory only, or a mix of both. By using temp files only, the cache size wouldn't have any noticeable impact on the memory footprint.

**Comment by ben-manes** [comment]
> My understanding seems to be different that yours. What the doc says is that the weight is not used to decide which entry is removed first, this will still be decided by the policy, like last accessed, or last written. However, it doesn't say that the cache entry wouldn't be removed if the size is over the maximum.

That is correct. The cache will load the entry and return it to the caller, but it may be immediately eligible for removal by an eviction policy. In your case if the entry's weight exceeds the maximum then the cache will discard it in preference removing it and clearing the cache. Obviously retaining what we can if preferable, as the new entry can't be held regardless.

Otherwise the eviction decision is based on recency and frequency. A future version will likely incorporate the weight in order to make a smarter decision and increase hit rates. This was explored in the paper [Lightweight Robust Size Aware Cache Management](https://arxiv.org/abs/2105.08770). That improvement offers a modest boost, but is a bit harder when considering concurrency and has not yet been raised by users as a concern.

> I checked the PDDocument docs, and it looks like you can supply MemoryUsageSettings which would allow you to restrict the amount of data you load into memory, and stores the rest in temporary files.

In my own usage of PdfBox (unrelated to Komga), I had to use mixed memory usage settings and disable the resource cache. The resource cache is soft reference based and causes OOME as images are humongous objects in G1's terminology, and GCs do not (or did not at the time) handle those well. If I recall correctly, the resource cache is document-specific so the likelihood of cache hits was negligible in my use-case (convert pdfs from or to images). The mixed setting was less helpful than I had hoped, but moving my work to AWS Lambda was much better thanks to isolated failures, minimized risk of resource exhaustion (serves only one request per instance), and cheaply scalable. Unfortunately we moved to Google CloudRun which offers only an in-memory file system, so MemoryUsageSetting lost its benefit. PdfBox also generates very bloated files, so I had to post process using ghostscript to avoid outbound emails from failing due to attachment limits. Eventually for that usage I'll stop dragging that original code forward and replace it all with mupdf-tools.

**Comment by gotson** [comment]
Thanks @ben-manes for that very detailed response ! Also appreciate your insights on Apache PDFBox.

I wonder how you ended up aware of that discussion here though 😅

**Comment by ben-manes** [comment]
Oh, GitHub search is global so a few keywords surfaces new public issues mentioning caffeine. Then if I might be able to help resolve confusion or catch a bug I’ll reply. It’s useful feedback to know what might need improvements.

**Comment by gotson** [comment]
> Oh, GitHub search is global so a few keywords surfaces new public issues mentioning caffeine. Then if I might be able to help resolve confusion or catch a bug I’ll reply. It’s useful feedback to know what might need improvements.

I'm really impressed, and big thanks for making Caffeine, i use it in a few projects and really like it :)

**Comment by gotson** [comment]
> I have tested to load PDFs using `PDDocument.load(path.toFile(), MemoryUsageSetting.setupTempFileOnly())`, this seems to be working.

@svew would you be able to test this on your pdf library, see if the impact on the ram is lower?

**Comment by svew** [comment]
Yes thanks again @ben-manes for clearing that up! I got the wrong idea from what I read 😅 An explicit note on behavior for oversized cache entries might be useful in the documentation!

> @svew would you be able to test this on your pdf library, see if the impact on the ram is lower?

Sure, I'm running Komga from docker image, and I'm a bit unfamiliar with using docker, how would I create and load a custom image of Komga? 

Also, do you notice any significant perf change when only using temp files versus using only memory? Because a hybrid solution might perform a bit better, only loading in up to 10 or 20MB or a PDF at a time would allow you to store a few pages in memory to serve while a user is reading. If we really wanted to optimize memory, we could have user loaded PDFs load some pages into memory, while analysis worker threads only ever load to temp files?

**Comment by svew** [comment]
And about the question about the reason to use a cache, my thinking was that between worker threads queuing jobs for book analysis, and a user reading a book one page at a time, it would make more sense to have the user sessions and worker threads to load the PDFs on their own, and close them when their sessions are over.

But of course, it's not always possible to know when a user session ends, so caching books for user sessions makes sense, and if you're already caching books for user sessions you could also use that cache for worker threads. It might be more optimal to detach worker threads from the cache and have them load and close PDDocuments on their own since it's not very likely they'll be reading the same book as the user? Unless multiple jobs for a single book could be queued at once and handled by multiple different workers...

It'd of course be better memory management-wise to close PDF documents immediately after you know you're finished using them, but I might be trying to solve problems that won't really exist once we start offloading the bulk of PDF document buffers to file :)

**Comment by svew** [comment]
Another thought, is it possible a thread might, at the beginning of its job, grab a `PDDocument` from cache, hold it for longer than the one minute expiry, and try to read the PDF after the cache has already closed the document? I haven't noticed this myself to be clear, but it does seem like a possibility unless the PDF document is fetched from the `PdfExtractor` every time a thread wants to use it

**Comment by gotson** [comment]
> Sure, I'm running Komga from docker image, and I'm a bit unfamiliar with using docker, how would I create and load a custom image of Komga?

You don't necessarily have to, you could test locally on your dev machine, with a test library. I need to update the instructions on how to create the Docker image.

> Also, do you notice any significant perf change when only using temp files versus using only memory?

Slight, but the impact is much less noticeable compared to removing the cache. The parsing really takes a toll.

> Because a hybrid solution might perform a bit better, only loading in up to 10 or 20MB or a PDF at a time would allow you to store a few pages in memory to serve while a user is reading.

That would really depend on the kind of PDF. Most PDFs are actually single image pages, and some are actual layered PDFs. I believe the single image PDF would not work well with the mixed mode, for instance if each image size is over the max memory allowed, they would go to disk anyway. And again, too many knobs would bring complexity, and more people asking for support.

> If we really wanted to optimize memory, we could have user loaded PDFs load some pages into memory, while analysis worker threads only ever load to temp files?

That would mean change the interface for extracting images (which is common to all the different formats), and thus leaking some implementation details into the interface. I think it's probably overkill at this stage.

> And about the question about the reason to use a cache, my thinking was that between worker threads queuing jobs for book analysis, and a user reading a book one page at a time, it would make more sense to have the user sessions and worker threads to load the PDFs on their own, and close them when their sessions are over.
> 
> But of course, it's not always possible to know when a user session ends, so caching books for user sessions makes sense, and if you're already caching books for user sessions you could also use that cache for worker threads. It might be more optimal to detach worker threads from the cache and have them load and close PDDocuments on their own since it's not very likely they'll be reading the same book as the user? Unless multiple jobs for a single book could be queued at once and handled by multiple different workers...
> 
> It'd of course be better memory management-wise to close PDF documents immediately after you know you're finished using them, but I might be trying to solve problems that won't really exist once we start offloading the bulk of PDF document buffers to file :)

The thing is that analysis only happens once, so optimizing for that is probably not worth it.

> Another thought, is it possible a thread might, at the beginning of its job, grab a `PDDocument` from cache, hold it for longer than the one minute expiry, and try to read the PDF after the cache has already closed the document? I haven't noticed this myself to be clear, but it does seem like a possibility unless the PDF document is fetched from the `PdfExtractor` every time a thread wants to use it

Unlikely, extracting a single page from the document will take a couple seconds max.

**Comment by gotson** [comment]
Closed by #785 and 8da2489671b4a1f00675b126abe71f6f2f57ea49
